File: 2023/src/bin/day3/main.rs
#![warn(clippy::pedantic)]
use std::{borrow::Cow, collections::HashMap, ops::Range};

use advent_2023::{BasicSolution, Solution};
use itertools::Itertools;
use winnow::{
    ascii::dec_uint,
    combinator::{iterator, preceded},
    stream::AsChar,
    token::take_till0,
    Located, Parser,
};

struct Day;

#[derive(Debug, Clone)]
struct Schematic<'a> {
    raw: Vec<&'a [u8]>,
    numbers: Vec<Number>,
}

#[derive(Debug, Clone)]
struct Number {
    row: usize,
    columns: Range<usize>,
    value: u32,
}

impl BasicSolution for Day {
    const INPUT: &'static str = include_str!("data.txt");
    const SAMPLE_INPUT: &'static str = include_str!("sample.txt");

    type Shared = Schematic<'static>;
    type Answer = u32;

    const SAMPLE_ANSWER_A: Self::TestAnswer = 4361;
    const SAMPLE_ANSWER_B: Self::TestAnswer = 467_835;

    fn shared(input: &'static str) -> anyhow::Result<Self::Shared> {
        let mut numbers = Vec::new();
        for (row, line) in input.lines().enumerate() {
            let mut iter_nums =
                iterator(Located::new(line), preceded(not_dec, dec_uint.with_span()));
            numbers.extend(iter_nums.map(|(value, columns)| Number {
                row,
                columns,
                value,
            }));
            // Purely for error checking
            let (rest, ()) = iter_nums.finish().map_err(anyhow::Error::msg)?;
            not_dec.parse(rest).map_err(anyhow::Error::msg)?;
        }
        let raw = input.lines().map(str::as_bytes).collect_vec();
        Ok(Schematic { raw, numbers })
    }

    fn part_a(schematic: Cow<Self::Shared>) -> anyhow::Result<Self::Answer> {
        Ok(schematic
            .numbers
            .iter()
            .filter(|number| {
                number.adjacent_coords().any(|coords| {
                    get_2d(&schematic.raw, coords).is_some_and(|c| !c.is_ascii_digit() && c != b'.')
                })
            })
            .map(|number| number.value)
            .sum())
    }

    fn part_b(Schematic { numbers, raw }: Self::Shared) -> anyhow::Result<Self::Answer> {
        let mut potential_gears: HashMap<(usize, usize), Vec<u32>> = HashMap::new();
        for num in numbers {
            num.adjacent_coords()
                .filter(|&coords| get_2d(&raw, coords) == Some(b'*'))
                .for_each(|coords| {
                    potential_gears.entry(coords).or_default().push(num.value);
                });
        }
        Ok(potential_gears
            .into_values()
            .filter(|values| values.len() == 2)
            .map(|values| values.into_iter().product::<u32>())
            .sum())
    }
}

fn not_dec(input: &mut Located<&'static str>) -> winnow::PResult<&'static str> {
    take_till0(AsChar::is_dec_digit).parse_next(input)
}

impl Number {
    fn adjacent_coords(&self) -> impl Iterator<Item = (usize, usize)> + '_ {
        let rows_to_check = self.row.saturating_sub(1)..=self.row + 1;
        let columns_to_check = self.columns.start.saturating_sub(1)..=self.columns.end;
        rows_to_check.cartesian_product(columns_to_check)
    }
}

fn get_2d(grid: &[&[u8]], (row, col): (usize, usize)) -> Option<u8> {
    grid.get(row).and_then(|line| line.get(col)).copied()
}

fn main() -> anyhow::Result<()> {
    Day::main()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a() -> anyhow::Result<()> {
        Day::test_part_a()
    }

    #[test]
    fn b() -> anyhow::Result<()> {
        Day::test_part_b()
    }
}

File: 2023/python/day3/main.py
import re
from collections import defaultdict
from dataclasses import dataclass
from itertools import chain
from pathlib import Path

SAMPLE_INPUT = """467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598..
"""
SAMPLE_INPUT_B = SAMPLE_INPUT
INPUT = Path("data.txt").read_text()

NUMBER = re.compile(r"\d+")


@dataclass
class Number:
    value: int
    row: int
    start_column: int
    end_column: int

    def adjacent_coords(self):
        return (
            (row, column)
            for row in range(self.row - 1, self.row + 2)
            for column in range(self.start_column - 1, self.end_column + 1)
        )


def part_a(input: str):
    grid = input.splitlines()
    numbers = parse_numbers(grid)
    return sum(
        number.value
        for number in numbers
        if any(
            ((c := get_2d(grid, row, column)) and c not in ".0123456789")
            for row, column in number.adjacent_coords()
        )
    )


def part_b(input: str):
    grid = input.splitlines()
    numbers = parse_numbers(grid)
    possible_gears = defaultdict(list)
    for number in numbers:
        for row, column in number.adjacent_coords():
            if get_2d(grid, row, column) == "*":
                possible_gears[(row, column)].append(number.value)
    return sum(vals[0] * vals[1] for vals in possible_gears.values() if len(vals) == 2)


def parse_numbers(grid: list[str]):
    return chain.from_iterable(
        find_numbers(line_num, line) for line_num, line in enumerate(grid)
    )


def find_numbers(line_num: int, line: str):
    return [
        Number(
            value=int(match.group(0)),
            row=line_num,
            start_column=match.start(),
            end_column=match.end(),
        )
        for match in NUMBER.finditer(line)
    ]


def get_2d(grid: list[str], row: int, column: int):
    if row < 0 or row >= len(grid):
        return None
    if column < 0 or column >= len(grid[row]):
        return None
    return grid[row][column]


def test_part_a(capsys):
    with capsys.disabled():
        assert part_a(SAMPLE_INPUT) == 4361
        print(f"Part A: {part_a(INPUT)}")


def test_part_b(capsys):
    assert part_b(SAMPLE_INPUT_B) == 467835
    with capsys.disabled():
        print(f"Part B: {part_b(INPUT)}")
